enhancements/ingress/global-admission-hsts.md

```diff
@@ -0,0 +1,728 @@
+        case DefaultIncludeSubdomains, NoOpinionIncludeSubdomains:
+        // anything is allowed, do nothing
+        case RequireIncludeSubdomains:
+            if !c.includeSubdomains {
+                return fmt.Errorf("includeSubdomains must be true")
+            }
+        case RequireNotIncludeSubdomains:
+            if c.includeSubdomains {
+                return fmt.Errorf("includeSubdomains must be unspecified")
+            }
+    }
+
+    return nil
+}
+// Check if the route matches the required domain/namespace in the HSTS Policy
+func requiredHSTSMatchesRoute(requirement RequiredHSTSPolicy, route *routeapi.Route, namespace *corev1.Namespace) (bool, error) {
+    matchesNamespace, err := matchesNamespaceSelector(requirement.NamespaceSelector, namespace)
+    if err != nil {
+        return false, err
+    }
+    if !matchesNamespace {
+        return false, nil
+    }
+    routeDomains := []string{route.Spec.Host}
+    for _, ingress := range route.Status.Ingress {
+        routeDomains = append(routeDomains, ingress.Host)
+    }
+    if matchesDomain(requirement.DomainPatterns, routeDomains) {
+        return true, nil
+    }
+
+    return false, nil
+}
+````
+
+Formatting of the HSTS Annotation field on a `route`:
+```yaml
+max-age=<numberofseconds>[;includeSubDomains][;preload]`
+```
+
+The `RequiredHSTSPolicies` variable can be configured by the administrator in the `Ingress`, as shown in this example, where
+there is one HSTS Policy is for domain abc.com, expires after one year, includes subdomains, and requests preload:
+```yaml
+apiVersion: config.openshift.io/v1
+kind: Ingress
+metadata:
+  name: cluster
+spec:
+  domain: apps.abc.com
+  requiredHSTSPolicies:
+    - domainPatterns:
+        - abc.com
+      maxAge:
+          smallestMaxAge: 1
+          largestMaxAge: 31536000
+      preloadPolicy: "RequirePreload"
+      includeSubdomainsPolicy: "RequireIncludeSubdomains"
+```
+
+#### Route Administration
+
+To handle upgraded clusters with non-compliant HSTS routes, the best solution is to update
+the manifests at the source and apply the updates.  However, in upgrade clusters, administrators
+can update the routes directly via the API.
+
+To apply HSTS to all routes in the cluster, use a command like this. (This command requires
+an updated API that supports the `--all-namespaces` flag on the `oc annotate` verb):
+```shell
+$ oc annotate route --all --all-namespaces --overwrite=true "haproxy.router.openshift.io/hsts_header"="max-age=31536000"
+```
+To apply HSTS to all routes in a particular namespace, use a command like this:
+```shell
+$ oc annotate route --all -n my-namespace --overwrite=true "haproxy.router.openshift.io/hsts_header"="max-age=31536000"
+```
+To remove HSTS from all routes in a particular namespace, set the `max-age` to 0, using a command like this:
+```shell
+$ oc annotate route --all -n my-namespace --overwrite=true "haproxy.router.openshift.io/hsts_header"="max-age=0"
```

Comment:
Have a look at the annotate removal syntax.  It allows a leading or trailing `-` to indicate removal.  I don't recall which.

Reply:
Removing the annotation is not the same as setting the max-age to 0.  The header will remain on the client until it's max-age times out, so setting the max-age to 0 is the correct way to remove the header.

Reply:
> Removing the annotation is not the same as setting the max-age to 0. The header will remain on the client until it's max-age times out, so setting the max-age to 0 is the correct way to remove the header.

Thanks for the clarification.  Can you add that explanation to the line above?